---
title: "sc_protocol_basic"
output: html_document
---

The following material provides an overview of a basic analysis of single cell transcriptome data.  The material is largely derived (sometimes verbatim) from the following sources:

1.  Seurat2:  http://satijalab.org/seurat/  
2.  Pagoda2:  https://github.com/hms-dbmi/pagoda2
3.  Shekhar et al. https://github.com/broadinstitute/BipolarCell2016

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library('pagoda2')
library(Matrix)
library(colorRamps)
library(dplyr)

# globals used below for organism-specific settings (human vs. mouse)
get_goSets_func = NULL
get_go_env = NULL
org = NULL
```

# Data set choice:
```{r}

##################################################
## ********* USER DEFINED SECTION ***************
##################################################

## Data set options:   un-comment just one of thhe 'counts_matrix_filename' options.

##########################################
# **  human peripheral blood mononuclear cells from: https://support.10xgenomics.com/single-cell-gene-expression/datasets
#counts_matrix_filename = "data/pbmcs/pbmc3k.counts.matrix.gz"; org="human"

##############################
# ** mouse retinal bipolar cells, from Shekhar et al. Cell 2016
counts_matrix_filename = "data/retinal_bipolar_cells/retinal_bipolar.dat.gz"; org="mouse"

```


```{r}
## Pagoda variable settings - depend on organism type
if (org == "human") {
  get_goSets_func = p2.generate.human.go.web
  get_go_env = p2.generate.human.go
  suppressMessages(library(org.Hs.eg.db))
  ALIAS2EG = org.Hs.egALIAS2EG
} else if (org == "mouse") {
  get_goSets_func = p2.generate.mouse.go.web
  get_go_env = pagoda2:::p2.generate.mouse.go
  suppressMessages(library(org.Mm.eg.db))
  ALIAS2EG = org.Mm.egALIAS2EG
} else {
  stop("Error, not sure what organism we're using")
}
```

##############################
# Data preparation
##############################

```{r}
# Read data from your file, rows as genes colums as cells
myCountMatrix <- read.table(gzfile(counts_matrix_filename), header=T, row.names=1)
```

Look at the matrix:
```{r}
myCountMatrix[1:10, 1:3]
```

How big is the matrix?

```{r}
dim(myCountMatrix) # report num rows and cols
```

Size in bytes?
```{r}
object.size(myCountMatrix)
```

Convert the matrix to a sparse matrix
```{r}
myCountMatrixSparse <- Matrix(as.matrix(myCountMatrix), sparse = T)

# take a look at it:
myCountMatrixSparse[1:10,1:3]
```

```{r}

# check dimensions:
dim(myCountMatrixSparse)

# check size:
object.size(myCountMatrixSparse)

# size reduction:
object.size(myCountMatrixSparse) / object.size(myCountMatrix)

```



```{r}
# Remove the original matrix to reduce memory usage
rm(myCountMatrix)
myCountMatrixSparse.prefiltered = myCountMatrixSparse # store just in case
```

## Filtering 'bad' cells

Look at the summary counts

```{r}
#par(mfrow=c(1,2), mar = c(3.5,3.5,2.0,0.5), mgp = c(2,0.65,0), cex = 1.0)

reads_per_cell = Matrix::colSums(myCountMatrixSparse)
reads_per_gene = Matrix::rowSums(myCountMatrixSparse)
genes_per_cell = Matrix::colSums(myCountMatrixSparse>0) # count gene only if it has non-zero reads mapped.
cells_per_gene = Matrix::rowSums(myCountMatrixSparse>0) # only count cells where the gene is expressed

hist(log10(reads_per_cell+1),main='reads per cell',col='wheat')
hist(log10(genes_per_cell+1), main='genes per cell', col='wheat')
plot(reads_per_cell, genes_per_cell, log='xy', col='wheat')
hist(log10(reads_per_gene+1),main='reads per gene',col='wheat')
```

### Plot genes per cell with cells ranked accordingly.

```{r}
plot(sort(genes_per_cell), xlab='cell', log='y', main='genes per cell (ordered)')

```

### Cell filtering criteria:  define min and max genes per cell

```{r}
##################################################
## ********* USER DEFINED SECTION ***************
##################################################

#  set upper and lower thresholds for genes per cell:
MIN_GENES_PER_CELL = 350  ## user-defined setting
MAX_GENES_PER_CELL = 1800  ## user-defined setting

# now replot with the thresholds being shown:
plot(sort(genes_per_cell), xlab='cell', log='y', main='genes per cell (ordered)')
abline(h=MIN_GENES_PER_CELL, col='green')  # lower threshold
abline(h=MAX_GENES_PER_CELL, col='green') # upper threshold
```


### Examine percent mitochondrial read content

```{r}
# define the mitochondrial genes
mito_genes = grep("^mt-", rownames(myCountMatrixSparse) , ignore.case=T, value=T)
print(mito_genes)
```



```{r}
# compute pct mito
mito_gene_read_counts = Matrix::colSums(myCountMatrixSparse[mito_genes,])
pct_mito = mito_gene_read_counts / reads_per_cell * 100
plot(sort(pct_mito))
```

Decide on maximum allowed percent mitochondrial reads:

```{r}
##################################################
## ********* USER DEFINED SECTION ***************
##################################################

MAX_PCT_MITO = 10   ## user-defined setting

plot(sort(pct_mito))
abline(h=MAX_PCT_MITO, col='red')

```


## cell selection as per Peter Karchenko - the Pagoda way

```{r}
df = data.frame(reads_per_cell=reads_per_cell, genes_per_cell=genes_per_cell)
head(df)
```

### Plot gene_per_cell vs. reads_per_cell, define outliers

```{r}
library(MASS)
df = df[order(df$reads_per_cell),] # order by reads_per_cell
plot(df, log='xy')
m <- rlm(genes_per_cell~reads_per_cell,data=df) # robust linear model, not sens to outliers
p.level = 1e-3
# predict genes_per_cell based on observed reads_per_cell
suppressWarnings(pb <- data.frame(predict(m, interval='prediction', 
                                          level = 1-p.level, # define conf interval
                                          type="response")))
polygon(c(df$reads_per_cell, rev(df$reads_per_cell)),
        c(pb$lwr, rev(pb$upr)), col=adjustcolor(2,alpha=0.1), border = NA)

# identifier outliers as having observed genes_per_cell outside the prediction confidence interval
outliers <- rownames(df)[df$genes_per_cell > pb$upr | df$genes_per_cell < pb$lwr];
points(df[outliers,],col=2,cex=0.6)
```


### Now, actually filter out 'bad' cells (and genes)

```{r}
myCountMatrixSparse = myCountMatrixSparse.prefiltered # just in case we re-run this block using different thresholds.

###############################################################
# prune genes, require a gene to be expressed in at least 3 cells

myCountMatrixSparse.prefiltered = myCountMatrixSparse
myCountMatrixSparse = myCountMatrixSparse[cells_per_gene >= 3,]  ## user can change this if needed.

###############################################################
# prune cells
valid_cells = colnames(myCountMatrixSparse) # all cells
message('starting with: ', length(valid_cells), ' cells') # number starting with

## remove cells based on gene count criteria:
valid_cells = valid_cells[genes_per_cell >= MIN_GENES_PER_CELL & genes_per_cell <= MAX_GENES_PER_CELL]  # set values based on your evaluation above
message('after filtering low and high gene count outliers: ', length(valid_cells), ' cells') # number after filtering based gene count thresholds

## remove cells having excessive mito read content
valid_cells = valid_cells[valid_cells %in% names(pct_mito)[pct_mito <= MAX_PCT_MITO]]
message('after removing high-mito cells: ', length(valid_cells), ' cells') # number remaining after high-mito cells removed

## remove cells identified as outliers via the Karchenko method
valid_cells = valid_cells[ ! valid_cells %in% outliers]
message('after removing final outliers: ', length(valid_cells), ' cells') # number surviving outlier detection

## update the count matrix to contain only the valid cells
myCountMatrixSparse = myCountMatrixSparse[,valid_cells]
```


#########################################
# Seurat2:  http://satijalab.org/seurat/
#########################################


```{r}
library(Seurat)
seurat2obj = CreateSeuratObject(raw.data=myCountMatrixSparse)
```

###  examine contents of seurat2obj
```{r}
str(seurat2obj)
```



```{r}
VlnPlot(object = seurat2obj, features.plot = c("nGene"), group.by = c('orig.ident'))
```

Can add metadata about cells: 

```{r}
# let's just recompute percent mito again based on our filtered matrix:
mito.genes <- grep(pattern = "^mt-", x = rownames(x = seurat2obj@raw.data), ignore.case = TRUE, value = TRUE)
percent.mito <- Matrix::colSums(seurat2obj@raw.data[mito.genes, ]) /
                                        Matrix::colSums(seurat2obj@raw.data)

# AddMetaData adds columns to object@data.info, and is a great place to stash QC stats
seurat2obj <- AddMetaData(object = seurat2obj, 
                          metadata = percent.mito, 
                          col.name = "percent.mito")

VlnPlot(object = seurat2obj, features.plot = c("nGene", "nUMI", "percent.mito"), nCol = 3)
```


### Normalizing the data

After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method "LogNormalize" that normalizes the gene expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. 


```{r normalize}
seurat2obj <- NormalizeData(object = seurat2obj, 
                            normalization.method = "LogNormalize", 
                            scale.factor = 1e4)
```

### Detection of variable genes across the single cells

Seurat calculates highly variable genes and focuses on these for downstream analysis. **`FindVariableGenes`** calculates the average expression and dispersion for each gene, places these genes into bins, and then calculates a z-score for dispersion within each bin. This helps control for the relationship between variability and average expression. This function is unchanged from (Macosko *et al*.), but new methods for variable gene expression identification are coming soon. We suggest that users set these parameters to mark visual outliers on the dispersion plot, but the exact parameter settings may vary based on the data type, heterogeneity in the sample, and normalization strategy. The parameters here identify ~2,000 variable genes, and represent typical parameter settings for UMI data that is normalized to a total of 1e4 molecules.

```{r var_genes}
seurat2obj <- FindVariableGenes(object = seurat2obj, mean.function = ExpMean, dispersion.function = LogVMR, x.low.cutoff = 0.2, x.high.cutoff = 8, y.cutoff = 1.0, num.bin=20)  # if this fails, experiment with the num.bin setting
```

```{r len_var_genes}
length(x = seurat2obj@var.genes)
```

```{r pca_pre_regress}
seurat2obj <- ScaleData(object = seurat2obj)
seurat2obj <- RunPCA(object = seurat2obj, pc.genes = seurat2obj@var.genes, do.print = TRUE, pcs.print = 1:2, genes.print = 5, pcs.compute = 40, maxit = 500, weight.by.var = FALSE)
PCAPlot(object = seurat2obj, dim.1 = 1, dim.2 = 2)
```

```{r}
FeaturePlot(seurat2obj, dim.1=1, dim.2=2, reduction.use='pca', features.plot=c('nGene'))
```



```{r saveobject}
save(seurat2obj, file = "seurat2obj-pre_batch_correct.Robj")
```


How many dimensions are there really?

```{r}
PCElbowPlot(object = seurat2obj)
```

```{r tsne}
seurat2obj <- RunTSNE(object = seurat2obj, dims.use = 1:10, do.fast = TRUE)
TSNEPlot(object = seurat2obj)
```



```{r}

## set up batch correction

if (org == 'mouse') {
  
  #########################################################################
  ## in our mouse retinal bipolar cell data set, we have batch information.
  
  # here we have to batches:
  #     batch 1 = { Bipolar 1, 2, 3, 4}
  #     batch 2 = { Bipolar 5, 6}
  
  batchname = seurat2obj@meta.data$orig.ident
  batchid = rep(1,length(batchname))
  batchid[batchname=="Bipolar5"] = 2
  batchid[batchname=="Bipolar6"] = 2
  names(batchid) = rownames(seurat2obj@meta.data)
  seurat2obj <- AddMetaData(object = seurat2obj, metadata = batchid, col.name = "batchid")
  table(seurat2obj@meta.data$batchid)
  
  # regress out the nGene, percent.mito, and batchid effects
  seurat2obj <- ScaleData(object = seurat2obj, vars.to.regress = c("nGene", "percent.mito", "batchid")) 
  seurat2obj <- RunPCA(object = seurat2obj, pc.genes = seurat2obj@var.genes, 
                       do.print = FALSE, pcs.compute = 40, weight.by.var = FALSE)
  
  # redo PCA and tSNE
  seurat2obj <- RunTSNE(object = seurat2obj, dims.use = 1:10, do.fast = TRUE)
  TSNEPlot(object = seurat2obj)
  
  # plot pca according to number of genes
  FeaturePlot(seurat2obj, dim.1=1, dim.2=2, reduction.use='tsne', features.plot=c('nGene', 'percent.mito'))

} else {
  
  ##############################################################
  ## Standard regressing out the nGene and percent.mito effects:
  
  # regress out the nGene effects
  seurat2obj <- ScaleData(object = seurat2obj, vars.to.regress = c("nGene", "percent.mito"))
  
  # rerun PCA on the regressed-out, 'cleaner' data
  seurat2obj <- RunPCA(object = seurat2obj, pc.genes = seurat2obj@var.genes, 
                       do.print = FALSE, pcs.compute = 40, weight.by.var = FALSE)
  
  # redo PCA and tSNE
  seurat2obj <- RunTSNE(object = seurat2obj, 
                        dims.use = 1:10, # pca dimensions to use
                        seed.use = 12345, # random seed, layout will differ on diff settings
                        do.fast = TRUE) # runs Barnes-hut t-SNE
  TSNEPlot(object = seurat2obj)
  
  # plot pca according to number of genes
  FeaturePlot(seurat2obj, dim.1=1, dim.2=2, reduction.use='tsne', 
              features.plot=c('nGene', 'percent.mito'))
  
}

```


## Cluster the cells

>Note, most text in this section is lifted verbatim from the official Seurat2 tutorial. Modifications to code and/or text aim to make it generic for use in just about any target data set.

Seurat now includes a graph-based clustering approach compared to (Macosko *et al*.). Importantly, the *distance metric* which drives the clustering analysis (based on previously identified PCs) remains the same. However, our approach to partioning the cellular distance matrix into clusters has dramatically improved. Our approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNA-seq data [[SNN-Cliq, Xu and Su, Bioinformatics, 2015]](http://bioinformatics.oxfordjournals.org/content/early/2015/02/10/bioinformatics.btv088.abstract) and CyTOF data [[PhenoGraph, Levine *et al*., Cell, 2015]](http://www.ncbi.nlm.nih.gov/pubmed/26095251). Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar gene expression patterns, and then attempt to partition this graph into highly interconnected 'quasi-cliques' or 'communities'. As in PhenoGraph, we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard distance). To cluster the cells, we apply modularity optimization techniques [[SLM, Blondel *et al*., Journal of Statistical Mechanics]](https://doi.org/10.1088/1742-5468/2008/10/P10008), to iteratively group cells together, with the goal of optimizing the standard modularity function.

The `FindClusters` function implements the procedure, and contains a resolution parameter that sets the 'granularity' of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.6-1.2 typically returns good results for single cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters are saved in the `object@ident` slot.


```{r cluster}
## - Find clusters ##

# save.SNN = T saves the SNN so that the clustering algorithm 
#           can be rerun using the same graph
# but with a different resolution value (see docs for full details)
seurat2obj <- FindClusters(object = seurat2obj, reduction.type = "pca", 
                           dims.use = 1:10, resolution = 0.8, 
                           print.output = 0, save.SNN = TRUE)

# note that you can set do.label=T to help label individual clusters
TSNEPlot(object = seurat2obj, do.label=T)
```


You can save the object at this point so that it can easily be loaded back in without having to rerun the computationally intensive steps performed above, or easily shared with collaborators.
```{r}
save(seurat2obj, file = "seurat2obj-post_batch_correct.Robj")
```

### Finding differentially expressed genes (cluster biomarkers)

Seurat can help you find markers that define clusters via differential expression. By default, it identifes positive and negative markers of a single cluster (specified in `ident.1`), compared to all other cells.  `FindAllMarkers` automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.

The min.pct argument requires a gene to be detected at a minimum percentage in either of the two groups of cells, and the thresh.test argument requires a gene to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of genes that are unlikely to be highly discriminatory. As another option to speed up these computations, max.cells.per.ident can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significiant and the most highly differentially expressed genes will likely still rise to the top.


The 'test.use' parameter indicates which test to use. Seurat currently implements "bimod" (likelihood-ratio test for single cell gene expression, McDavid et al., Bioinformatics, 2013, default), "roc" (standard AUC classifier), "t" (Students t-test), and "tobit" (Tobit-test for differential gene expression, as in Trapnell et al., Nature Biotech, 2014), 'poisson', and 'negbinom'. The latter two options should only be used on UMI datasets, and assume an underlying poisson or negative-binomial distribution.


```{r}

##################################################
## ********* USER DEFINED SECTION ***************
##################################################

MIN_LOGFOLD_CHANGE = 1 # set to minimum required average log fold change in gene expression.
MIN_PCT_CELLS_EXPR_GENE = 0.1  # minimum percent of cells that must express gene in either clstr.


all.markers = FindAllMarkers(seurat2obj, 
                             min.pct = MIN_PCT_CELLS_EXPR_GENE,
                             thresh.use = MIN_LOGFOLD_CHANGE, 
                             only.pos = TRUE,
                             test.use="bimod") # likelihood ratio test
```




```{r}
# sort all the markers by p-value
all.markers.sortedByPval = all.markers[order(all.markers$p_val),]

# take a look at the top most significant markers
head(all.markers.sortedByPval)
```

# make a heatmap showing the top 10 markers

```{r}
top10 <- all.markers.sortedByPval %>%  group_by(cluster)  %>% do(head(., n=10))
DoHeatmap(object = seurat2obj, genes.use = top10$gene, slim.col.label = TRUE, remove.key = TRUE)
```



```{r}
# examine the top 4 markers in the context of the tSNE plots:
FeaturePlot(seurat2obj, features.plot = all.markers.sortedByPval$gene[1:4])
```



### Get genes uniquely DE in each cluster:
```{r}
genes_uniquely_DE = all.markers.sortedByPval %>% dplyr::filter(avg_diff > MIN_LOGFOLD_CHANGE) %>% group_by(gene) %>%  summarize(n=n()) %>%  filter(n==1)

genes_uniquely_DE.markers.sortedByPval = 
  all.markers.sortedByPval[all.markers.sortedByPval$gene 
                           %in% genes_uniquely_DE$gene,]

top_marker_each = genes_uniquely_DE.markers.sortedByPval %>% 
      dplyr::filter(avg_diff >= MIN_LOGFOLD_CHANGE) %>% 
      dplyr::group_by(cluster) %>% 
      do(head(., n=1))  # set to higher value to get that number of top markers each clst.
print(top_marker_each)
```




```{r}
for (gene in top_marker_each$gene) {
  FeaturePlot(seurat2obj, features.plot = gene)
}

```

### Other ways to explore marker genes:
Violin plots - examine distribution of gene expression across clustered cells.

```{r}
for (i in 1:length(top_marker_each$gene)) {
  print(VlnPlot(seurat2obj, features.plot = top_marker_each$gene[i]))
}
```

The DotPlot: a concise way of exploring percent of cells expressing a gene and the gene expression intensity.

```{r}
DotPlot(seurat2obj, genes.plot=unique(top_marker_each$gene))
```



***
## Merging of expression proximal clusters

>Note, derived verbatim from seurat2 tutorials.

In the section below, we explore an overclustering combined with post-hoc merging strategy that can help discover weaker splits in the data
This section applies cutoffs in an admittedly supervised way, and is something we are actively working to improve
As a more conservative approach, ignore the section below, and use a resolution value of 2 as shown above.

We can bump the resolution up to call more clusters, but this slightly over-clusters the data

In fact there is no 'perfect' value of resolution, we always either under or over-cluster the data
This is because of dramatically different cluster sizes, and is known as the 'multi-resolution' problem in graph-based clustering
One solution is to slightly over-cluster the data, and then perform a post-hoc merging step, where transcriptionally indistinguishable clusters are merged back together

As a test for merging, we use the Out-of-bag error (OOBE) from a random forest classifier, but you could also set a cutoff for # of differentially expressed genes

### Build classification hierarchy

Build a classification hierarchy that places transcriptionally similar clusters adjacent on a tree

```{r}
seurat2obj <- BuildClusterTree(seurat2obj, do.reorder = TRUE, reorder.numeric = TRUE)
TSNEPlot(object = seurat2obj, do.label = TRUE) # tSNE plot as a reference to the cluster ids
```

### Run RandomForest predictor, compute OOBE

Calculate the classification error for left/right cells on each branch of the tree
sort internal nodes based on OOBE. For nodes with high OOBE, we cannot accurately tell the left/right children apart based on random forests, so the clusters may need to be merged



```{r}
node.scores <- AssessNodes(seurat2obj)
node.scores = node.scores[order(node.scores$oobe, decreasing = TRUE), ] 
```


```{r}
print(head(node.scores))
```


```{r}
save(seurat2obj, file = "seurat2obj.pre-cluster-merge.Robj")
```

```{r}

##################################################
## ********* USER DEFINED SECTION ***************
##################################################

max_allowed_oobe = 0.10   # any tree branches exceeding this OOBE are collapsed.

## must re-play this section for each merge event, since only one merge per 'play'

node_to_merge = node.scores$node[node.scores$oobe > max_allowed_oobe]
if (length(node_to_merge) > 0) {
  # have at least one node to merge
  message("merging high OOBE clusters")
  # do merge for top set:
  seurat2obj = MergeNode(object=seurat2obj, node.use=node_to_merge[1])
  
  ####################################################################################
  # now redo the few steps we performed just above - in this new post-merge clustering
  
  # examine the new tree post-merge
  seurat2obj <- BuildClusterTree(seurat2obj, do.reorder = TRUE, reorder.numeric = TRUE)
  TSNEPlot(object = seurat2obj, do.label = TRUE) # tSNE plot as a reference to the cluster ids
  
  # reexamine classification error in the context of the updated clusters:
  node.scores <- AssessNodes(seurat2obj)
  node.scores = node.scores[order(node.scores$oobe, decreasing = TRUE), ]
  print(head(node.scores))
} else {
  message("no clusters with high OOBE to merge.")
}
```

Now you can go back and redefine differentially expressed transcripts based on these new cluster identities.

## ------------ End of Seurat2 tutorial -------------------------------------##
## Visit http://satijalab.org/seurat/  for many more tutorials and vignettes ##
## --------------------------------------------------------------------------##



#######################################################################
# Pagoda2 - hierarchical differential expression and gene set analysis  
#######################################################################


```{r}

# Generate a new pagoda2 object
myPagoda2Object <- Pagoda2$new(x = myCountMatrixSparse, n.cores = 2)

# Adjust the variance
myPagoda2Object$adjustVariance(plot = T, gam.k = 10)

# Calculate a PCA reduction with the number of PCs specified by nPCs
# and using only the n.odgenes overdispersed genes -- in this case 2000
myPagoda2Object$calculatePcaReduction(nPcs = 100, n.odgenes = 2.e3)

# Generate K-nearest neighbour graph
myPagoda2Object$makeKnnGraph(k = 20, type = 'PCA', center = T,
    weight.type = 'none', n.cores = 2, distance = 'cosine')

```

##############################
# Identify clusters via Pagoda 
##############################

infomap: "Infomap is a network clustering algorithm based on the Map equation." (https://github.com/mapequation/infomap)

multilevel.community: "Fast unfolding of community hierarchies in large networks, http://arxiv.org/abs/arXiv:0803.0476 for the details."

walktrap.community: "This function is the implementation of the Walktrap community finding algorithm, see Pascal Pons, Matthieu Latapy: Computing communities in large networks using random walks, http://arxiv.org/abs/physics/0512106"

>Each of the above is supported by the R igraph package: http://igraph.org/r/


```{r}

# Identify clusters using the infomap.community method
# on the basis of the reduction called 'PCA' (generated above)
# Save the resulting clustering as 'infomap'
myPagoda2Object$getKnnClusters(method = infomap.community,
                               type = 'PCA', name = 'infomap')
                               
# Do an independent identification of clusters using the
# multilevel community algorithm again using the PCA reduction
# and save it as 'multilevel'. This does not overwrite the 
# previous clustering
myPagoda2Object$getKnnClusters(method = multilevel.community,
                               type = 'PCA', name='multilevel')
                               
# Do yet another clustering
myPagoda2Object$getKnnClusters(method = walktrap.community,
                               type = 'PCA', name='walktrap')

```


#########################################
# Generate embeddings: largeVis and tSNE
#########################################

Embedding type support:

t-SNE: "t-Distributed Stochastic Neighbor Embedding (t-SNE) is a (prize-winning) technique for dimensionality reduction that is particularly well suited for the visualization of high-dimensional datasets." (https://lvdmaaten.github.io/tsne/)

LargeVis: "LargeVis, a technique that first constructs an accurately approximated K-nearest neighbor graph from the data and then layouts the graph in the low-dimensional space. Comparing to t-SNE, LargeVis significantly reduces the computational cost of the graph construction step and employs a principled probabilistic model for the visualization step, the objective of which can be effectively optimized through asynchronous stochastic gradient descent with a linear time complexity. The whole procedure thus easily scales to millions of high-dimensional data points. " (https://arxiv.org/abs/1602.00370)


```{r}
# Generate an embedding with largeVis on the basis of the PCA reduction
M <- 30 
myPagoda2Object$getEmbedding(
  type = 'PCA',
  embeddingType = 'largeVis',
  M = M,
  perplexity = 30, 
  gamma = 1 / M,
  alpha = 1
)

# Generate an embedding with tSNE on the basis of the PCA
# reduction
myPagoda2Object$getEmbedding(type = 'PCA', 
                             embeddingType = 'tSNE', 
                             perplexity = 30)  
# if above errors due to perplexity being too high, set it to a lower value.

```

####################################################
# Plot the generated embeddings and clustering types
####################################################

You can explore the results of different embedding and clustering methods.  Note, that 2D embedding is just for visualization, as the clustering methods do not use the embeddings but instead operate on the multidimensional data directly (usually after dimensionality reduction).


```{r}
# largeVis & infomap
myPagoda2Object$plotEmbedding(type = 'PCA',
                              embedding = 'largeVis',
                              mark.clusters = T,
                              clusterType = 'infomap')
                              
# largeVis & multilevel          
myPagoda2Object$plotEmbedding(type = 'PCA',
                              embedding = 'largeVis',
                              mark.clusters = T,
                              clusterType = 'multilevel')

# largeVis & walktrap
myPagoda2Object$plotEmbedding(type = 'PCA',
                              embedding = 'largeVis',
                              mark.clusters = T,
                              clusterType = 'walktrap')

# infomap & tSNE
myPagoda2Object$plotEmbedding(type = 'PCA',
                              embeddingType = 'tSNE',
                              clusterType = 'infomap',
                              mark.clusters = T)
```


##########################################################################
# Differential Gene expression: pagoda2-hierarchical marker identification
##########################################################################


```{r}
# Pathway overdispersion -- required for web
go.env <- get_go_env(myPagoda2Object)

myPagoda2Object$testPathwayOverdispersion(setenv = go.env, verbose = T, 
                                          correlation.distance.threshold = 0.9, # so, cor >= (1-0.9) get merged.
                                          recalculate.pca = F,
                                          min.pathway.size = 100, max.pathway.size = 1000)



# Calculate the differential gene expression of each cluster
# defined in multilevel clustering on the basis of the PCA reduction (uses Wilcoxon test)
myPagoda2Object$getDifferentialGenes(type='PCA', clusterType = 'multilevel')

# Plot a differential expression heatmap using the differentially expressed genes
# above
#myPagoda2Object$plotDiffGeneHeatmap(type='PCA', clusterType = 'multilevel')

# can access the DE genes via:  myPagoda2Object$diffgenes

## try to set up for pagoda2-hierarchical marker identification

# translate gene names to ids
ids <- unlist(lapply(mget(colnames(myPagoda2Object$counts), ALIAS2EG, ifnotfound=NA), function(x) x[1]))
# reverse map
rids <- names(ids); names(rids) <- ids;
# list all the ids per GO category
go.env <- list2env(eapply(ALIAS2EG, function(x) as.character(na.omit(rids[x]))))

myPagoda2Object$testPathwayOverdispersion(go.env, verbose=T, 
                                          correlation.distance.threshold=0.95, 
                                          recalculate.pca=F,
                                          top.aspects=15)

myPagoda2Object$getHierarchicalDiffExpressionAspects(type='PCA', clusterName='walktrap', z.threshold=3)
```


## Generate a Pagoda-1 web-app

```{r}
appHierarchDE <- p2.make.pagoda1.app(myPagoda2Object, 
                            inner.clustering=TRUE, 
                            embeddingType='tSNE',
                            clusterType='walktrap', 
                            min.group.size=50,
                            row.clustering=list(order=rev(1:nrow(myPagoda2Object$misc$pathwayOD$xv))))
```

## Launch Pagoda-1 web-app
```{r}
show.app(appHierarchDE, 'p2hierarchical', browse=T)
```

###############################################
## Gene set (Gene Ontology) dispersion analysis
###############################################


```{r}
################################
# Generate the web application
################################

# Generate GO genesets for the web app
myGeneNames <- colnames(myPagoda2Object$counts)
goSets <- get_goSets_func(myGeneNames)

# Generate differental expression between each cluster and everything else
# Load these clusters as pre-defined gene sets with the given prefix
deSets <- get.de.geneset(myPagoda2Object, groups = myPagoda2Object$clusters$PCA[[1]], prefix = 'de_')

# Merge Genesets
geneSets <- c(goSets, deSets)

# Additional metadata generation
additionalMetadata <- list()

additionalMetadata$altCluster <- p2.metadata.from.factor(myPagoda2Object$clusters$PCA[[1]], 
                                                         displayname = 'Infomap', 
                                                         s = 0.7, v = 0.8,start = 0, end = 0.5)

additionalMetadata$altCluster2 <- p2.metadata.from.factor(myPagoda2Object$clusters$PCA[[2]], 
                                                         displayname = 'Multilevel', 
                                                         s = 0.7, v = 0.8,start = 0, end = 0.5)


additionalMetadata$altCluster3 <- p2.metadata.from.factor(myPagoda2Object$clusters$PCA[[3]], 
                                                          displayname = 'Walktrap')

# Generate the gene knn graph, which is used to find gene with similar expression patterns

myPagoda2Object$makeGeneKnnGraph()

# Generate and display web app
myPagoda2WebObject <-
  make.p2.app(
    myPagoda2Object,
    dendrogramCellGroups = myPagoda2Object$clusters$PCA[[1]],
    additionalMetadata = additionalMetadata,
    geneSets = geneSets,
    show.clusters = FALSE, # Hide the clusters that were used for the dendrogram from the metadata
  )

```

####################################
## App Loading #####################
####################################


```{r}  
show.app(app=myPagoda2WebObject,name='p2GOdispersion')
```



